File: Maze/src/mazePD/Droid.java
package mazePD;

import java.util.HashSet;
import java.util.function.IntSupplier;

import mazePD.Maze.Content;
import mazePD.Maze.Direction;
import mazeStack.LinkedStack;

public class Droid implements DroidInterface {

    /** Called after each committed move (not during probe/undo). */
    @FunctionalInterface
    public interface StepListener {
        void onStep(Coordinates pos, int stepCount, boolean backtrack);
    }

    /** Called once when exploration finishes. */
    @FunctionalInterface
    public interface CompleteListener {
        void onComplete(boolean found, int totalSteps);
    }

    // Direction order matching scanAdjLoc() indices: 0=D00, 1=D90, 2=D180, 3=D270
    private static final Direction[] SCAN_DIRS =
        { Direction.D00, Direction.D90, Direction.D180, Direction.D270 };

    private String name;
    private Coordinates currentLocation;
    private final LinkedStack<Coordinates> stack   = new LinkedStack<>();
    private final HashSet<Coordinates>     visited = new HashSet<>();

    private StepListener     stepListener;
    private CompleteListener completeListener;
    private IntSupplier      delaySupplier;

    public Droid(String name) { this.name = name; }

    @Override
    public String getName() { return this.name; }

    public void setStepListener(StepListener l)        { this.stepListener    = l; }
    public void setCompleteListener(CompleteListener l) { this.completeListener = l; }
    /** Supply a delay in milliseconds between each committed move. */
    public void setStepDelay(IntSupplier supplier)     { this.delaySupplier   = supplier; }

    public void exploreMaze(Maze maze) {
        int stepCount = 0;

        Coordinates start = maze.enterMaze(this);
        stack.push(start);
        visited.add(start);
        setCurrentLocation(start);
        fireStep(start, stepCount, false);
        sleep();

        while (!stack.isEmpty() && !maze.scanCurLoc(this).equals(Content.END)) {
            if (Thread.currentThread().isInterrupted()) return;

            Content[]   adj  = maze.scanAdjLoc(this);
            Coordinates next = getNextMove(adj, maze);
            stepCount++;

            if (next != null) {
                stack.push(next);
                visited.add(next);
                fireStep(next, stepCount, false);
                sleep();

                // Use portal if standing on one
                if (maze.scanCurLoc(this).equals(Content.PORTAL_DN)) {
                    Coordinates dest = maze.usePortal(this, Direction.DN);
                    stack.push(dest);
                    visited.add(dest);
                    stepCount++;
                    fireStep(dest, stepCount, false);
                    moveToTopStack(maze);
                    sleep();
                }
            } else {
                stack.pop();
                if (!stack.isEmpty()) {
                    moveToTopStack(maze);
                    fireStep(stack.peek(), stepCount, true);
                    sleep();
                }
            }
        }

        boolean found = !stack.isEmpty() && maze.scanCurLoc(this).equals(Content.END);
        if (completeListener != null) completeListener.onComplete(found, stepCount);
        if (!found) System.out.println("no path found");
    }

    /**
     * Scans adjacent cells and moves the droid to the first unvisited
     * traversable neighbour, returning its coordinates.
     * Probe moves that land on an already-visited cell are immediately reversed.
     * Returns null if every direction is blocked or already visited.
     */
    public Coordinates getNextMove(Content[] mazeContent, Maze maze) {
        for (int i = 0; i < mazeContent.length; i++) {
            Content c = mazeContent[i];
            if (c == Content.EMPTY || c == Content.PORTAL_DN
                    || c == Content.PORTAL_UP || c == Content.END) {

                Coordinates candidate = maze.move(this, SCAN_DIRS[i]);
                if (!isInVisited(candidate)) {
                    return candidate;       // droid is now at candidate
                }
                // Undo the probe: reverse direction index is (i+2)%4
                maze.move(this, SCAN_DIRS[(i + 2) % 4]);
            }
        }
        return null;
    }

    /** Moves the droid one step toward the current top of the stack. */
    public void moveToTopStack(Maze maze) {
        if (stack.isEmpty()) return;
        int x = maze.getCurrentCoordinates(this).x;
        int y = maze.getCurrentCoordinates(this).y;
        Coordinates top = stack.peek();
        if (top.x == x) {
            maze.move(this, top.y > y ? Direction.D180 : Direction.D00);
        } else if (top.y == y) {
            maze.move(this, top.x > x ? Direction.D90 : Direction.D270);
        }
    }

    /** O(1) visited check — relies on Coordinates.hashCode() and equals(). */
    public boolean isInVisited(Coordinates coord) {
        return visited.contains(coord);
    }

    public Coordinates getCurrentLocation() { return currentLocation; }
    public void setCurrentLocation(Coordinates c) { this.currentLocation = c; }

    @Override
    public String toString() {
        return "Path Taken: " + stack.toArrayFromLast();
    }

    // ── Private helpers ───────────────────────────────────────

    private void fireStep(Coordinates pos, int step, boolean backtrack) {
        if (stepListener != null) stepListener.onStep(pos, step, backtrack);
    }

    private void sleep() {
        int delay = (delaySupplier != null) ? delaySupplier.getAsInt() : 0;
        if (delay <= 0) return;
        try {
            Thread.sleep(delay);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: web/droid.js
/**
 * Droid — step-by-step DFS solver, ported from Droid.java.
 *
 * Call step() once per animation frame to advance one move.
 * Reads from a Maze instance; never modifies it.
 */
class Droid {
  /** @param {Maze} maze */
  constructor(maze) {
    this.maze      = maze;
    this.pos       = { ...maze.start };          // current position
    this.stack     = [{ ...maze.start }];        // DFS path stack
    this.visited   = new Set([_key(maze.start)]);// all visited cells (O(1) lookup)
    this.pathSet   = new Set([_key(maze.start)]);// cells currently on the stack
    this.stepCount = 0;
    this.done      = false;
    this.found     = false;

    // Direction vectors matching Java scanAdjLoc() order: N, E, S, W
    this._dirs = [
      { dx:  0, dy: -1 },  // D00  — north (decrease y)
      { dx:  1, dy:  0 },  // D90  — east  (increase x)
      { dx:  0, dy:  1 },  // D180 — south (increase y)
      { dx: -1, dy:  0 },  // D270 — west  (decrease x)
    ];
  }

  /**
   * Advance one step of DFS exploration.
   * @returns {boolean} true while exploration is still running, false when done.
   */
  step() {
    if (this.done) return false;
    if (this.stack.length === 0) { this.done = true; return false; }

    const { x, y } = this.stack[this.stack.length - 1]; // peek

    // Reached the exit?
    if (this.maze.getContent(x, y) === 'END') {
      this.done  = true;
      this.found = true;
      this.pos   = { x, y };
      return false;
    }

    // Try each direction in order (matching Java's D00, D90, D180, D270)
    let moved = false;
    for (const { dx, dy } of this._dirs) {
      const nx  = x + dx;
      const ny  = y + dy;
      const key = _key({ x: nx, y: ny });

      if (this.maze.canOccupy(nx, ny) && !this.visited.has(key)) {
        this.stack.push({ x: nx, y: ny });
        this.visited.add(key);
        this.pathSet.add(key);
        this.pos = { x: nx, y: ny };
        this.stepCount++;
        moved = true;
        break;
      }
    }

    if (!moved) {
      // Dead end — backtrack: pop current cell off the path
      const popped = this.stack.pop();
      this.pathSet.delete(_key(popped));
      this.stepCount++;

      if (this.stack.length === 0) {
        this.done = true;
        return false;
      }
      // Move droid back to new top of stack
      this.pos = { ...this.stack[this.stack.length - 1] };
    }

    return !this.done;
  }
}

/** Canonical string key for a {x, y} coordinate. */
function _key({ x, y }) { return `${x},${y}`; }
